datafusion/physical-plan/src/sorts/progressive_eval.rs

```diff
@@ -0,0 +1,563 @@
+            }
+
+            let input_stream = spawn_buffered(streams.remove(0), 1);
+            num_read_inputs_counter.add(1);
+
+            if i == 0 {
+                current_input_stream = Some(input_stream);
+            } else {
+                prefetched_input_streams.push(input_stream);
+            }
+        }
+
+        Ok(Self {
+            input_stream_count,
+            num_input_streams_to_prefetch,
+            current_stream_idx,
+            current_input_stream,
+            prefetched_input_streams,
+            num_read_inputs_counter,
+            streams,
+        })
+    }
+
+    /// Set next available stream to current_input_stream
+    /// Also prefetch one more input stream if not all of them are prefetched yet
+    fn next_stream(&mut self) {
+        // No more input stream
+        if self.current_stream_idx >= self.input_stream_count {
+            // panic if we have not reached the end of all input streams
+            assert!(
+                self.prefetched_input_streams.is_empty(),
+                "Internal error in ProgressiveEvalStream: There should not have input streams left to read",);
+
+            self.current_input_stream = None;
+        } else {
+            // prefetch one more input stream before setting next strem to the current input stream
+            if self.current_stream_idx + self.num_input_streams_to_prefetch
+                < self.input_stream_count
+            {
+                self.num_read_inputs_counter.add(1);
+                self.prefetched_input_streams
+                    .push(spawn_buffered(self.streams.remove(0), 1));
+            }
+
+            self.current_stream_idx += 1;
+            if self.prefetched_input_streams.is_empty() {
+                self.current_input_stream = None;
+            } else {
+                self.current_input_stream = Some(self.prefetched_input_streams.remove(0));
+            }
+        }
+    }
+
+    fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<Result<RecordBatch>>> {
+        // All input streams have been read
+        if self.current_input_stream.is_none() {
+            return Poll::Ready(None);
+        }
+
+        // Get next record batch
+        let mut poll;
+        loop {
+            poll = self
+                .current_input_stream
+                .as_mut()
+                .unwrap()
+                .poll_next_unpin(cx);
+            match poll {
+                // This input stream no longer has data, move to next stream
+                Poll::Ready(None) => {
+                    self.next_stream();
+                    if self.current_input_stream.is_none() {
+                        // Have reached the end of all input streams
+                        return Poll::Ready(None);
+                    }
+                }
+                _ => break,
+            }
+        }
+
+        poll
+    }
+}
+
+/// Concat input streams until reaching the fetch limit
+struct ProgressiveEvalStream {
```

Comment:
Both of your stream structures are a close mirror to how we do it too. 😆

Reply:
https://github.com/apache/datafusion/pull/10490

Yes, from the PR